pkg/collect/utilisation.go

```diff
@@ -0,0 +1,363 @@
+}
+
+type Utilisation struct {
+	Readings map[string][]UsageReading `json:"readings,omitempty" yaml:"readings,omitempty"`
+}
+
+type CollectCPUMemUtilisation struct {
+	Collector    *troubleshootv1beta2.CPUMemUsage
+	BundlePath   string
+	Namespace    string
+	ClientConfig *rest.Config
+	Client       kubernetes.Interface
+	Context      context.Context
+	RBACErrors
+}
+
+func (c *CollectCPUMemUtilisation) Title() string {
+	return getCollectorName(c)
+}
+
+func (c *CollectCPUMemUtilisation) IsExcluded() (bool, error) {
+	return isExcluded(c.Collector.Exclude)
+}
+
+// todo: Rework.
+func (c *CollectCPUMemUtilisation) Collect(progressChan chan<- interface{}) (CollectorResult, error) {
+	thanosServiceNamespace := "kommander"
+	output := NewResult()
+
+	// Get the pods of the deployment
+	pods, err := c.Client.CoreV1().Pods(thanosServiceNamespace).List(context.Background(), metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/component=query,app.kubernetes.io/instance=thanos,app.kubernetes.io/name=thanos",
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not look up thanos-query pods: ")
+	}
+	if len(pods.Items) < 1 {
+		// Check if no thanos-query pods are found.
+		return nil, errors.New("Could not find any thanos-query pods.")
+	}
+	podName := pods.Items[0].Name
+
+	// Create a Prometheus API client using the API server endpoint
+	prometheusClient, stopCh, err := StartPortForwarding(context.Background(), c.ClientConfig, coreout.NewNonInteractiveShell(os.Stdout, os.Stderr, 4), podName)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error querying Prometheus for CPU utilization:")
+	}
+	v1api := v1.NewAPI(prometheusClient)
+
+	// Query for CPU utilization
+	cpuQuery := "sum(rate(container_cpu_usage_seconds_total{container!='POD', container!=''}[1h])) by (namespace)"
+	start := time.Now().Add(-1 * time.Duration(c.Collector.Duration) * 24 * time.Hour)
+	end := time.Now()
+	step := time.Duration(c.Collector.SamplingStep) * time.Hour
+	cpuResult, _, err := v1api.QueryRange(context.Background(), cpuQuery, v1.Range{
+		Start: start,
+		End:   end,
+		Step:  step,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "Error querying Prometheus for CPU utilization: ")
+	}
+
+	memoryQuery := "sum(container_memory_working_set_bytes{container!='POD', container!=''}) by (namespace)"
+	memoryResult, _, err := v1api.QueryRange(context.Background(), memoryQuery, v1.Range{
+		Start: start,
+		End:   end,
+		Step:  step,
+	})
+
+	if err != nil {
+		return nil, errors.Wrap(err, "Error querying Prometheus for memory utilization: ")
+	}
+	// Process query result
+	memMatrix, ok := memoryResult.(model.Matrix)
+	if !ok {
+		return nil, errors.New("Invalid query result, expected matrix")
+	}
+
+	// Process query result
+	cpuMatrix, ok := cpuResult.(model.Matrix)
+	if !ok {
+		return nil, errors.New("Invalid query result, expected matrix")
+	}
+	metricsByNamespace := make(map[string]map[string][]model.SamplePair, 0)
```

Comment:
is it not that the length of this map is: `len(cpuMatrix)+len(memMatrix)` right?